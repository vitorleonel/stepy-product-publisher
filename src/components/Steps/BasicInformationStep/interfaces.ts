import { CategoryItem } from '../CategoryStep/interfaces';

export interface ViewProps {
  category: CategoryItem;
}
