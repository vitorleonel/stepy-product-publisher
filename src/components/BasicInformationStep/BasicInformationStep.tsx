import { PanelProps, Button } from "@blueprintjs/core";

import ImagesStep from "../ImagesStep";

import { ViewProps } from "./interfaces";

const BasicInformationStep = (props: PanelProps<ViewProps>) => {
  const openPanel = () => props.openPanel({
    renderPanel: ImagesStep
  })

  return (
    <>
      <p>BasicInformationStep</p>

      <Button text="Next" onClick={openPanel} />
    </>
  );
}

export default BasicInformationStep;
