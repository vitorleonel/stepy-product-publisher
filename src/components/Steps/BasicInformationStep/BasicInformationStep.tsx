import { PanelProps, FormGroup, InputGroup, TextArea, NumericInput } from "@blueprintjs/core";
import { useEffect, useState } from "react";

import StepHeader from "../../StepHeader";
import StepNavigation from "../../StepNavigation";
import ImagesStep from "../ImagesStep";

import { ViewProps } from "./interfaces";

const BasicInformationStep = (props: PanelProps<ViewProps>) => {
  const [title, setTitle] = useState<string>('');
  const [description, setDescription] = useState<string>('');
  const [price, setPrice] = useState<number>(0);

  const [nextButtonIsDisabled, setNextButtonIsDisabled] = useState<boolean>(true);

  useEffect(() => {
    if(!title || !description || !price) {
      setNextButtonIsDisabled(true);
    } else {
      setNextButtonIsDisabled(false);
    }
  }, [title, description, price])

  const openPanel = () => props.openPanel({
    renderPanel: ImagesStep,
    title: 'ImagesStep',
    props: {
      ...props,
      title,
      description,
      price,
    },
  });

  const formatPrice = (price: number) => {
    const formatter = new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      maximumFractionDigits: 0,
      minimumFractionDigits: 0,
    });

    return formatter.format(price).replace('$', '');
  }

  const handlePrice = (_: number, currentPrice: string) => {
    const normalizedPrice = parseInt((currentPrice || '0').replace(/[^0-9]+/g, ''));

    setPrice(normalizedPrice);
  }

  return (
    <section className="steps-item">
      <StepHeader
        title="Basic product information"
        description="Take care in the title and description of the product as this is a very important factor when the user will decide for your product."
      />

      <FormGroup label="Title" labelFor="title-input">
        <InputGroup
          id="title-input"
          placeholder="Your product's title"
          large
          fill
          value={title}
          onChange={(event) => setTitle(event.target.value)}
        />
      </FormGroup>

      <FormGroup label="Description" labelFor="description-input">
        <TextArea
          id="description-input"
          placeholder="Your product's description"
          rows={5}
          large
          fill
          growVertically
          value={description}
          onChange={(event) => setDescription(event.target.value)}
        />
      </FormGroup>

      <FormGroup label="Price" labelFor="price-input">
        <NumericInput
          id="price-input"
          placeholder="Your product's price"
          leftIcon="dollar"
          min={0}
          buttonPosition="none"
          large
          fill
          value={formatPrice(price)}
          onValueChange={handlePrice}
        />
      </FormGroup>

      <StepNavigation
        nextText="Product images"
        nextDisabled={nextButtonIsDisabled}
        prevHandler={props.closePanel}
        nextHandler={openPanel}
      />
    </section>
  );
}

export default BasicInformationStep;
