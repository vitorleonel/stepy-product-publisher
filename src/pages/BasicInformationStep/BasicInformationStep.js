import { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';

import StepHeader from '../../components/StepHeader';
import StepNavigation from '../../components/StepNavigation';
import InputField from '../../components/InputField';
import TextAreaField from '../../components/TextAreaField';

import Actions from '../../store/actions';

const BasicInformationStep = ({ state, dispatch }) => {
  const history = useHistory();

  const [title, setTitle] = useState(state.title);
  const [description, setDescription] = useState(state.description);
  const [price, setPrice] = useState(state.price);

  const [nextButtonIsDisabled, setNextButtonIsDisabled] = useState(true);

  useEffect(() => {
    if (!title || !description || !price) {
      setNextButtonIsDisabled(true);
    } else {
      setNextButtonIsDisabled(false);
    }
  }, [title, description, price]);

  const handleSubmit = () => {
    dispatch({
      type: Actions.SET_DATA,
      payload: { title, description, price },
    });
    history.push('/images');
  };

  const formatPrice = (price) => {
    const formatter = new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      maximumFractionDigits: 0,
      minimumFractionDigits: 0,
    });

    return formatter.format(price).replace('$', '');
  };

  const handlePrice = (currentPrice) => {
    const normalizedPrice = parseInt(
      (currentPrice || '0').replace(/[^0-9]+/g, '')
    );

    setPrice(normalizedPrice);
  };

  return (
    <section className="steps-item">
      <StepHeader
        title="Basic information"
        description="Take care in the title and description of the product as this is a very important factor when the user will decide for your product."
      />

      <InputField
        name="title"
        label="Title"
        placeholder="Your product's title"
        id="title-input"
        value={title}
        onChange={(event) => setTitle(event.target.value)}
      />

      <TextAreaField
        rows={5}
        name="description"
        label="Description"
        placeholder="Your product's description"
        id="description-input"
        value={description}
        onChange={(event) => setDescription(event.target.value)}
      />

      <InputField
        name="price"
        label="Price"
        placeholder="Your product's price"
        id="price-input"
        value={formatPrice(price)}
        onChange={(event) => handlePrice(event.target.value)}
      />

      <StepNavigation
        nextText="Images"
        nextDisabled={nextButtonIsDisabled}
        prevHandler={history.goBack}
        nextHandler={handleSubmit}
      />
    </section>
  );
};

export default BasicInformationStep;
