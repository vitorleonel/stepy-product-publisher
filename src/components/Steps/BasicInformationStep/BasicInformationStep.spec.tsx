import { render, fireEvent } from "@testing-library/react";

import BasicInformationStep from "./BasicInformationStep";

describe('BasicInformationStep Component', () => {

  const fakeCategory = {
    id: 1,
    name: 'Category',
    description: 'Description',
  };

  test('should render correctly', async () => {
    const { container } = render(<BasicInformationStep
      closePanel={jest.fn()}
      openPanel={jest.fn()}
      category={fakeCategory}
    />);

    expect(container).toMatchSnapshot();
  });

  test('should be possible to fill in the fields and call openPanel function', async () => {
    const openPanel = jest.fn();

    const { getByPlaceholderText, getByTestId } = render(<BasicInformationStep
      closePanel={jest.fn()}
      openPanel={openPanel}
      category={fakeCategory}
    />);

    const nextButton = getByTestId('nextButton');
    const titleField = getByPlaceholderText("Your product's title");
    const descriptionField = getByPlaceholderText("Your product's description");
    const priceField = getByPlaceholderText("Your product's price");

    await fireEvent.change(titleField, { target: { value: 'Test' } });
    await fireEvent.change(descriptionField, { target: { value: 'Test' } });
    await fireEvent.change(priceField, { target: { value: '1000' } });
    await fireEvent.click(nextButton);

    expect(titleField).toHaveValue('Test');
    expect(descriptionField).toHaveValue('Test');
    expect(priceField).toHaveValue('1,000');
    expect(openPanel).toBeCalledTimes(1);
  });

  test('should not be possible call openPanel function when I not fill fields', async () => {
    const openPanel = jest.fn();

    const { getByTestId } = render(<BasicInformationStep
      closePanel={jest.fn()}
      openPanel={openPanel}
      category={fakeCategory}
    />);

    const nextButton = getByTestId('nextButton');
    await fireEvent.click(nextButton);

    expect(openPanel).toBeCalledTimes(0);
  });

});
