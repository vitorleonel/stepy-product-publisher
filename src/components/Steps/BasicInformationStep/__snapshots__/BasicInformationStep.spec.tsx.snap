// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicInformationStep Component should render correctly 1`] = `
<div>
  <section
    class="steps-item"
  >
    <div
      class="steps-item__header"
    >
      <h2
        class="bp3-heading"
      >
        Basic product information
      </h2>
      <p
        class="bp3-running-text bp3-text-large"
      >
        Take care in the title and description of the product as this is a very important factor when the user will decide for your product.
      </p>
    </div>
    <div
      class="bp3-form-group"
    >
      <label
        class="bp3-label"
        for="title-input"
      >
        Title
         
        <span
          class="bp3-text-muted"
        />
      </label>
      <div
        class="bp3-form-content"
      >
        <div
          class="bp3-input-group bp3-fill bp3-large"
        >
          <input
            class="bp3-input"
            id="title-input"
            placeholder="Your product's title"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="bp3-form-group"
    >
      <label
        class="bp3-label"
        for="description-input"
      >
        Description
         
        <span
          class="bp3-text-muted"
        />
      </label>
      <div
        class="bp3-form-content"
      >
        <textarea
          class="bp3-input bp3-fill bp3-large"
          id="description-input"
          placeholder="Your product's description"
          rows="5"
          style="height: 0px;"
        />
      </div>
    </div>
    <div
      class="bp3-form-group"
    >
      <label
        class="bp3-label"
        for="price-input"
      >
        Price
         
        <span
          class="bp3-text-muted"
        />
      </label>
      <div
        class="bp3-form-content"
      >
        <div
          class="bp3-control-group bp3-fill bp3-numeric-input bp3-large"
        >
          <div
            class="bp3-input-group bp3-large"
          >
            <span
              class="bp3-icon bp3-icon-dollar"
              icon="dollar"
            >
              <svg
                data-icon="dollar"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  dollar
                </desc>
                <path
                  d="M12.83 9.51c-.1-.3-.25-.58-.45-.84s-.45-.49-.75-.7c-.3-.2-.65-.36-1.05-.48-.16-.04-.43-.11-.8-.2-.35-.09-.73-.18-1.12-.28-.39-.1-.74-.19-1.06-.27-.31-.08-.49-.12-.54-.13-.43-.12-.78-.29-1.05-.52-.27-.23-.4-.55-.4-.95 0-.29.07-.53.21-.72.14-.19.32-.34.54-.46.22-.11.46-.19.72-.24.26-.05.52-.08.77-.08.74 0 1.35.15 1.83.46.48.3.75.83.81 1.56h2.14c0-.6-.13-1.13-.38-1.58-.25-.45-.59-.84-1.02-1.15-.43-.31-.93-.54-1.49-.7-.24-.06-.49-.1-.75-.14V1c0-.55-.45-1-1-1s-1 .45-1 1v1.08c-.23.03-.46.07-.68.13-.54.13-1.02.34-1.44.61-.42.28-.76.63-1.02 1.05-.26.43-.39.93-.39 1.5 0 .3.04.59.13.88.09.29.23.56.44.82.21.26.48.49.83.7.35.21.79.38 1.31.51.85.21 1.56.38 2.14.52.58.13 1.08.28 1.52.42.25.09.48.23.69.44.21.21.32.53.32.97 0 .21-.05.42-.14.63-.09.21-.24.39-.45.55-.21.16-.47.29-.81.39-.33.1-.73.15-1.2.15-.43 0-.84-.05-1.21-.14-.37-.09-.7-.24-.99-.43-.29-.2-.51-.45-.67-.76-.16-.31-.24-.68-.24-1.12H3c.01.71.15 1.32.43 1.84.27.52.64.94 1.1 1.27.46.33.99.58 1.61.74.27.07.56.12.85.16V15c0 .55.45 1 1 1s1-.45 1-1v-1.05c.3-.03.61-.08.9-.15.58-.13 1.1-.34 1.56-.63.46-.29.83-.66 1.11-1.11.28-.45.42-1 .42-1.64 0-.31-.05-.61-.15-.91z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <input
              autocomplete="off"
              class="bp3-input"
              id="price-input"
              min="0"
              placeholder="Your product's price"
              type="text"
              value="0"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="steps-item__navigation"
    >
      <button
        class="bp3-button bp3-large"
        data-testid="prevButton"
        type="button"
      >
        <span
          class="bp3-icon bp3-icon-arrow-left"
          icon="arrow-left"
        >
          <svg
            data-icon="arrow-left"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              arrow-left
            </desc>
            <path
              d="M13.99 6.99H4.41L7.7 3.7a1.003 1.003 0 00-1.42-1.42l-5 5a1.014 1.014 0 000 1.42l5 5a1.003 1.003 0 001.42-1.42L4.41 8.99H14c.55 0 1-.45 1-1s-.46-1-1.01-1z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="bp3-button-text"
        >
          Previous
        </span>
      </button>
      <button
        class="bp3-button bp3-disabled bp3-large bp3-intent-success"
        data-testid="nextButton"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Product images
        </span>
        <span
          class="bp3-icon bp3-icon-arrow-right"
          icon="arrow-right"
        >
          <svg
            data-icon="arrow-right"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              arrow-right
            </desc>
            <path
              d="M14.7 7.29l-5-5a.965.965 0 00-.71-.3 1.003 1.003 0 00-.71 1.71l3.29 3.29H1.99c-.55 0-1 .45-1 1s.45 1 1 1h9.59l-3.29 3.29a1.003 1.003 0 001.42 1.42l5-5c.18-.18.29-.43.29-.71s-.12-.52-.3-.7z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </section>
</div>
`;
